src/project.js

```diff
@@ -155,25 +167,28 @@ function percentify (value, limit) {
     return (value / limit) * 100;
 }
 
+// implementation of floydWarshall alg for calculating visibility matrix in O(n^3) instead of O(n^4) with successive raising of powers
 function createVisibilityMatrix (result) {
-    var product = result.adjacencyMatrix, sum = result.adjacencyMatrix, changeCost = 0, visibilityMatrix;
 
-    result.adjacencyMatrix.forEach(function () {
-        product = matrix.multiply(product, result.adjacencyMatrix);
-        sum = matrix.add(product, sum);
-    });
+    var changeCost = 0, visibilityMatrix = adjacencyToDistMatrix(result.adjacencyMatrix), matrixLen = visibilityMatrix.length, k, i, j;
 
-    result.adjacencyMatrix.forEach(function (ignore, index) {
-        sum[index][index] = 1;
-    });
+    for (k = 0; k < matrixLen; k += 1) {
+        for (i = 0; i < matrixLen; i += 1) {
+            for (j = 0; j < matrixLen; j += 1) {
+                visibilityMatrix[i][j] = Math.min(visibilityMatrix[i][j], visibilityMatrix[i][k] + visibilityMatrix[k][j]);
+            }
+        }
+    }
 
-    visibilityMatrix = sum.map(function (row, rowIndex) {
+    visibilityMatrix = visibilityMatrix.map(function (row, rowIndex) {
         return row.map(function (value, columnIndex) {
-            if (value > 0) {
+            if (value < Infinity) {
                 changeCost += 1;
 
                 if (columnIndex !== rowIndex) {
                     return 1;
+                } else {
```

Comment:
Sorry, I meant the `else` itself rather than the clause inside it. The contents of the `if` condition above always returns, so we don't need to say `else` because it is already implied. But anyway, I'm bike-shedding to the extreme at this point, sorry. :)